src/controller/CHIPDeviceController.cpp

```diff
@@ -1304,13 +1330,98 @@ void OnBasicFailure(void * context, CHIP_ERROR error)
+        // We do not need to reset the failsafe here because we want to keep everything on the device up to this point, so just
+        // send the completion callbacks.
+        CommissioningStageComplete(CHIP_NO_ERROR);
+        SendCommissioningCompleteCallbacks(nodeId, commissioningCompletionStatus);
+    }
+    else
+    {
+        // If we've failed somewhere in the early stages (or we don't have a failedStage specified), we need to start from the
+        // beginning. However, because some of the commands can only be sent once per arm-failsafe, we also need to force a reset on
+        // the failsafe so we can start fresh on the next attempt.
+        GeneralCommissioning::Commands::ArmFailSafe::Type request;
+        request.expiryLengthSeconds = 0; // Expire immediately.
+        request.breadcrumb          = 0;
+        ChipLogProgress(Controller, "Expiring failsafe on proxy %p", proxy);
+        mDeviceBeingCommissioned = proxy;
+        // We actually want to do the same thing on success or failure because we're already in a failure state
+        SendCommand<GeneralCommissioningCluster>(proxy, request, OnDisarmFailsafe, OnDisarmFailsafeFailure);
+    }
+}
+
+void DeviceCommissioner::OnDisarmFailsafe(void * context,
+                                          const GeneralCommissioning::Commands::ArmFailSafeResponse::DecodableType & data)
+{
+    ChipLogProgress(Controller, "Failsafe disarmed");
+    DeviceCommissioner * commissioner = static_cast<DeviceCommissioner *>(context);
+    commissioner->DisarmDone();
+}
+
+void DeviceCommissioner::OnDisarmFailsafeFailure(void * context, CHIP_ERROR error)
+{
+    ChipLogProgress(Controller, "Received failure response  when disarming failsafe%s\n", chip::ErrorStr(error));
+    DeviceCommissioner * commissioner = static_cast<DeviceCommissioner *>(context);
+    commissioner->DisarmDone();
+}
+
+void DeviceCommissioner::DisarmDone()
+{
+    // At this point, we also want to close off the pase session so we need to re-establish
+    CommissioneeDeviceProxy * commissionee = FindCommissioneeDevice(mDeviceBeingCommissioned->GetDeviceId());
+
+    // Signal completion - this will reset mDeviceBeingCommissioned.
+    CommissioningStageComplete(CHIP_NO_ERROR);
+    SendCommissioningCompleteCallbacks(commissionee->GetDeviceId(), commissioningCompletionStatus);
+
+    // If we've disarmed the failsafe, it's because we're starting again, so kill the pase connection.
+    if (commissionee != nullptr)
+    {
+        ReleaseCommissioneeDevice(commissionee);
+    }
+}
+
+void DeviceCommissioner::SendCommissioningCompleteCallbacks(NodeId nodeId, const CompletionStatus & completionStatus)
+{
+    mCommissioningStage = CommissioningStage::kSecurePairing;
+    if (mPairingDelegate == nullptr)
+    {
+        return;
+    }
+    mPairingDelegate->OnCommissioningComplete(nodeId, completionStatus.err);
+    PeerId peerId(GetCompressedFabricId(), nodeId);
```

Comment:
No, that won't work.  Having a scoped node ID implies having a fabric table, and the commissioner may not have one if it is delegating operational network rendezvous to another node, which is what happens in the split commissioner / admin topology.

Reply:
Ah, apologies.  So I am slightly incorrect.  The node that sends commissioning complete definitely will have established CASE, so my comment is not necessarily applicable: this isn't the split commissioner / admin topology in this particular code path.

However, there is still the larger point that we need to be careful not to assume the commissioner has a fabric table.  It may not.

Reply:
I'm still not totally clear where ScopedNodeId is appropriate as opposed to PeerId TBH. I opted for PeerId here on the assumption that when we're talking about commissioning, we want to talk about commissioning-specific information, and the compressed fabric ID is appropriate in that context. It also matches to the Status callback, where the ScopedNodeId  seems inappropriate because commissioning is ongoing.

Reply:
> A PeerId is not a unique identifier, in general, because a compressed fabric id is not unique. In any situation where there might be multiple root public keys involved we should not use PeerId, because those might collide.

If we are uncomfortable with possible collisions, we need a structure that holds the root public key.  There are use cases in commissioning where a fabric table won't exist.

Reply:
https://github.com/project-chip/connectedhomeip/issues/16994